### PR TITLE
patch for setting install_name during install

### LINKED ISSUE #22: investigate setting install name
on macOS, dylibs should have their install_name (set with `install_name_tool -id`) set to the name that should be used to load the library. At its *most* specific, this should be the ABI-compatible version (`*.4.dylib` or `*.4.5.dylib`). However, the libs from suitesparse appear to set their install_name to their *full* name (e.g. `libsuitesparseconfig.4.5.4.dylib`), indicating that every patch-level release can make ABI changes.

This means that all downstream packages have to pin their dependency on suitesparse to the patch-level, because any upgrade to suitesparse, even a compatible one, will result in an `image not found` error on the old location:

```
Library not loaded: @rpath/libsuitesparseconfig.4.5.3.dylib
  Referenced from: /Users/minrk/conda/envs/petsc/lib/libslepc.3.7.dylib
  Reason: image not found
```

(here slepc shows an example of how this can be done: libslepc's install_name is `libslepc.3.7.dylib`, even though the actual file is `libslepc.3.7.1.dylib`)

Setting the install_name to the general (version-less) path or the ABI-compatibility version would allow for upgrading suitesparse to a compatible version without needing to rebuild the downstream package.

This should probably be reported upstream to suitesparse itself, but it can be addressed in the conda package by running `install_name_tool -id ...` in build.sh after `make install`.

### PR DESCRIPTION
**Pull request by minrk**
I've submitted this patch upstream. Not sure yet if it will be accepted.

closes #22

### CODE REVIEW

**Comment by basnijholt** [comment]
LGTM. 

@jakirkham @grlee77 what do you think?

**Comment by grlee77** [comment]
I am not familiar enough with `Makefile` syntax to judge if this is correct, so I would defer to those with more expertise.  Two quick questions:

1.) In short, this patch only modifies behavior on OS X and causes the dynamic libraries to use e.g.  `*.4.5.dylib` rather than the more specific `*.4.5.4.dylib` to be used as the `install_name`?

2.) Do we know that `SuiteSparse` has ABI compatibility between patch versions?

**Comment by minrk** [comment]
> In short, this patch only modifies behavior on OS X and causes the dynamic libraries to use e.g. *.4.5.dylib rather than the more specific *.4.5.4.dylib to be used as the install_name?

According to the variables in the Makefiles, it would be `*.4.dylib` (major only), but otherwise yes.

> Do we know that SuiteSparse has ABI compatibility between patch versions?

We know that suitesparse is setting the compatibility version in its dylibs, so at least the libraries themselves *claim* compatibility at the major version level. I can't speak for the project in general, but a cursory test of a conda package compiled against 4.5.3 updated manually to load 4.5.4 does work.

**Comment by grlee77** [comment]
sorry this PR has gone neglected.  I think now the build number should be bumped to 202 since the OpenBLAS update bumped it to 201.  

@minrk:  Is this otherwise good to go?  if interested, please add yourself as a maintainer as well

**Comment by jakirkham** [comment]
According to the [ABI report for SuiteSparse]( https://abi-laboratory.pro/tracker/timeline/suitesparse/ ) it appears that at least 4.5 has been stable across patch release. Before that it was static libraries that are fairly old, which makes it hard to say much about its compatibility.

**Comment by jakirkham** [comment]
Should we add a file test to ensure that the library name is as expected?

**Comment by minrk** [comment]
Good idea, I'll add a few calls to `otool -L` that verify the install name is right.

Aside: it would be cool if conda-build test environments could verify compatibility by running the test env with multiple versions of the built package.

In general, a useful pattern for verifying C libs that their linking config is set up correctly:

1. conda-build the lib
2. conda-build a *test* package that links the lib
3. install and test the test package against the oldest and newest versions of the lib that its dependency matrix would allow

**Comment by minrk** [comment]
Sorry for taking so long with the tests. The tests are here now, so we can start getting a suitesparse with correct linking on mac.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -10,10 +10,13 @@ source:
   fn: SuiteSparse-{{ version }}.tar.gz
   url: http://faculty.cse.tamu.edu/davis/SuiteSparse/SuiteSparse-{{ version }}.tar.gz
   sha256: de5fb496bdc029e55955e05d918a1862a177805fbbd5b957e8b5ce6632f6c77e
+  patches:
+    - install_name.patch
 
 build:
   skip: true  # [win]
   number: 200
```
Comment (isuruf) [inline recipe/meta.yaml]: Can you remove this line?

File: recipe/meta.yaml [hunk 2]
```diff
@@ -10,10 +10,13 @@ source:
...
+  number: 201
   features:
     - blas_{{ variant }}
```
Comment (isuruf) [inline recipe/meta.yaml]: Can you add 
```
run_export:
  - {{ pin_subpackage("suitesparse") }}
```
here?

**Comment by minrk** [comment]
This is now also a full conda-build 3 update per @isuruf's request.

**Comment by minrk** [comment]
Rebased again

**Comment by minrk** [comment]
I think this is good to go now (and has been for a year or so). Anything else blocking the merge?

**Comment by SylvainCorlay** [comment]
:+1:

**Comment by minrk** [comment]
Conflicts resolved again, build number bumped again.

**Comment by minrk** [comment]
All builds have succeeded. This should be ready to go again. Not sure why osx shows as still queued, since it's [succeeded a while ago](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=73578&view=logs).

**Pull request merged**